### PR TITLE
implement with fewer afoldl methods

### LINKED ISSUE #39175: inference regression on `afoldl`
I tried my big algebra package on 1.6.0-beta1. Many parts are slightly faster, but one important part is many times slower.
I found several problems but the shortest MWE is

```
struct Perm
  d::Vector{Int}
end
function Perm(x::Integer...)
  if isempty(x) return Perm(Int[]) end
  d=Int.(1:max(x...))
  for i in 1:length(x)-1
    d[x[i]]=x[i+1]
  end
  d[x[end]]=x[1]
  Perm(d)
end
struct SymmetricGroup
  gens::Vector{Perm}
end
SymmetricGroup(n::Int)=SymmetricGroup([Perm(i,i+1) for i in 1:n-1])

using BenchmarkTools

# on julia 1.5.3
@btime SymmetricGroup(20)
  1.080 μs (39 allocations: 3.72 KiB)

# on julia 1.6.0-beta1
@btime SymmetricGroup(20)
  7.658 μs (115 allocations: 6.39 KiB)
```

### PR DESCRIPTION
**Pull request by vtjnash**
With constant-propagation, inference (and Varargs runtime) is likely
better able to handle this version now (and it removes the n^2 behavior
definitions for semi-low argument counts).

Now we also need to force Vararg specialization up to 16 arguments
manually, so we do that explicitly (and slightly hack-y).

Closes #39175, as this fixes the regression, though it appears in my testing that https://github.com/JuliaLang/julia/pull/35675 also work very well here (permits inlining at a different layer, where it is more effective over a broader scope than just this regression)

### CODE REVIEW

File: base/operators.jl [hunk 1]
```diff
@@ -577,17 +577,37 @@ xor(x::Integer) = x
...
+    i =  7; y = op(y, bs[i]); l == i && return y
+    i =  8; y = op(y, bs[i]); l == i && return y
+    i =  9; y = op(y, bs[i]); l == i && return y
+    i = 10; y = op(y, bs[i]); l == i && return y
+    i = 11; y = op(y, bs[i]); l == i && return y
+    i = 12; y = op(y, bs[i]); l == i && return y
+    i = 13; y = op(y, bs[i]); l == i && return y
+    i = 14; y = op(y, bs[i]); l == i && return y
+    i = 15; y = op(y, bs[i]); l == i && return y
+    for i in (i + 1):length(bs)
```
Comment (JeffBezanson) [inline base/operators.jl:605]:
```suggestion
    for i in (i + 1):l
```

File: base/operators.jl [hunk 2]
```diff
@@ -577,17 +577,37 @@ xor(x::Integer) = x
...
+        y = op(y, bs[i])
+    end
+    return y
 end
+typeof(afoldl).name.mt.max_args = 18
```
Comment (JeffBezanson) [inline base/operators.jl:610]: :see_no_evil:

**Review by JeffBezanson — APPROVED** [review]

File: base/operators.jl [hunk 3]
```diff
@@ -577,17 +577,37 @@ xor(x::Integer) = x
...
-# foldl for argument lists. expand recursively up to a point, then
-# switch to a loop. this allows small cases like `a+b+c+d` to be inlined
+# foldl for argument lists. expand fully up to a point, then
+# switch to a loop. this allows small cases like `a+b+c+d` to be managed
 # efficiently, without a major slowdown for `+(x...)` when `x` is big.
-afoldl(op,a) = a
-afoldl(op,a,b) = op(a,b)
-afoldl(op,a,b,c...) = afoldl(op, op(a,b), c...)
-function afoldl(op,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,qs...)
-    y = op(op(op(op(op(op(op(op(op(op(op(op(op(op(op(a,b),c),d),e),f),g),h),i),j),k),l),m),n),o),p)
```
Comment (Sacha0) [inline base/operators.jl:587]: This line 😂

**Review by Sacha0 — APPROVED** [review]
Huh neat approach! :)

File: base/operators.jl (hunk 3 above)
Comment (JeffBezanson) [reply base/operators.jl:587]: I know; sad we have to drop one of my favorite lines of code :joy:

**Pull request merged**